keps/sig-api-machinery/3716-webhook-predicates/README.md

```diff
@@ -0,0 +1,857 @@
+  - [Rollout, Upgrade and Rollback Planning](#rollout-upgrade-and-rollback-planning)
+  - [Monitoring Requirements](#monitoring-requirements)
+  - [Dependencies](#dependencies)
+  - [Scalability](#scalability)
+  - [Troubleshooting](#troubleshooting)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+  - [Resource Exclusions](#resource-exclusions)
+  - [CEL Admission Control](#cel-admission-control)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests for meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes adding "predicate" expressions to admission webhooks, as an extension to the
+existing `rules` to define the scope of a webhook. A `predicate` is a
+[CEL](https://github.com/google/cel-spec) expression that must evaluate to true for the admission
+request to be sent to the webhook. If a `predicate` evaluates to false, the webhook is skipped for
+that request (implicitly allowed).
+
+<<[UNRESOLVED naming ]>>
+I'm not satisfied with the name `predicates` for this feature. Other ideas considered:
+- filters, prefilters
+- match{Predicates, Filters, Criteria, Expressions, Constraints, ...}
```

Comment:
Conditions?

Reply:
Yeah, over on https://github.com/kubernetes/enhancements/pull/3697#discussion_r1063011479 we landed on Preconditions as a good option. I might change this PR to use that instead.

Reply:
+1 to preconditions

Reply:
`matchExpressions` was my 1st choice, but I can get behind `preconditions` if that resonates with people

Reply:
FYI, we're also considering `matchConditions` on https://github.com/kubernetes/enhancements/pull/3697#discussion_r1063011479

Reply:
I pushed an update going with `matchConditions`, but we can still decide to change it.